Addressing #899 and #898.

jmintb opened the pull request:
Changes:
Added Subdirectory browsing
The previously selected folder is remembered.
The subdirectory is remembered for each Syncthing folder.
The saved subdirectory in the shared preferences is deleted when the corresponding folder is deleted from Syncthing.
The root directory of the FolderPickerActivity is set to the synced folder, so the user can only choose a sub folder within the folder that is being synced.
TheFolderPickerActivity was modified in order to allow for a custom root directory to be set.

Example with no sub folder selected:
![screenshot_20170821-141950](https://user-images.githubusercontent.com/10332534/29525503-2c9d2108-8693-11e7-8ff5-babc06707aee.png)

Example with a sub folder selected:
![screenshot_20170821-162521](https://user-images.githubusercontent.com/10332534/29525521-3fc21eb4-8693-11e7-8796-d72706a6d9ed.png)

AudriusButkevicius commented on src/main/java/com/nutomic/syncthingandroid/activities/FolderPickerActivity.java:
```diff
@@ -48,6 +49,9 @@
             "com.nutomic.syncthingandroid.activities.FolderPickerActivity.INITIAL_DIRECTORY";
 
+    public static final String EXTRA_ROOT_DIRECTORY =
+            "com.nutomic.syncthingandroid.activities.FolderPickerActivity.ROT_DIRECTORY";
```
ROT -> ROOT

AudriusButkevicius commented on src/main/java/com/nutomic/syncthingandroid/activities/FolderPickerActivity.java:
```diff
@@ -101,7 +116,8 @@ protected void onCreate(Bundle savedInstanceState) {
 
     /**
-     * Reads available storage devices/folders from various APIs and inserts them into
+     * If a root directory is specefied it is added to {@link #mRootsAdapter} otherwise
```
specefied -> specified

AudriusButkevicius commented on src/main/java/com/nutomic/syncthingandroid/activities/ShareActivity.java:
```diff
@@ -52,12 +65,24 @@ public void onApiChange(SyncthingService.State currentState) {
 
         List<Folder> folders = getApi().getFolders();
 
+        //Get the index of the previously selected folder.
```
Space after //

jmintb commented:
Thanks for the nitpicking :).

Nutomic commented on src/main/java/com/nutomic/syncthingandroid/activities/FolderActivity.java:
```diff
@@ -360,6 +361,9 @@ private Dialog createDeleteDialog(){
                     getApi().removeFolder(mFolder.id);
+                    //Remove saved data from share activity for this folder.
+                    SharedPreferences prefs = this.getSharedPreferences(ShareActivity.SHARED_PREFS_SAVED_DIRECTORIES, MODE_PRIVATE);
+                    prefs.edit().remove(mFolder.id).apply();
```
it's probably better to handle this in RestApi.removeFolder(), in case that function is called from somewhere else. And you dont need to set MODE_PRIVATE, because that's the default.

jmintb replied:
Agreed.

Nutomic commented on src/main/java/com/nutomic/syncthingandroid/activities/FolderPickerActivity.java:
```diff
@@ -74,6 +78,17 @@ public static Intent createIntent(Context context, String currentPath) {
+
+        if(!TextUtils.isEmpty(rootDir)) {
+            intent.putExtra(EXTRA_ROOT_DIRECTORY, rootDir);
+        }
```
If the function is called createIntentWithRootDir(), there is no reason to ever call it without a rootDir parameter. So I dont see the reason for this check.

jmintb replied:
I wanted to avoid an empty string "", but I guess that wouldn't actually be a problem.

Nutomic replied:
Yeah I don't think that's something you should handle really. Alternatively, you could have only one `createIntent()` method, and make the rootDir param nullable. That would also simplify the code a bit (one method less).

Nutomic commented on src/main/java/com/nutomic/syncthingandroid/activities/ShareActivity.java:
```diff
@@ -52,12 +65,24 @@ public void onApiChange(SyncthingService.State currentState) {
 
+        // Get the index of the previously selected folder.
+        int folderIndex = 0;
+        String savedFolderId = getSharedPreferences(SHARED_PREFS_SAVED_DIRECTORIES, MODE_PRIVATE)
```
Just use PreferenceManager.getDefaultSharedPreferences() instead.

jmintb replied:
In that case, should I use something like saved_sub_directory_(insert folder id here) as the key for the saved sub directories, instead of just the folder id. Just to ensure that the key is unique?

Nutomic replied:
Yeah that seems easier and less complicated.

Nutomic commented on src/main/java/com/nutomic/syncthingandroid/activities/ShareActivity.java:
```diff
@@ -123,7 +150,34 @@ protected void onCreate(Bundle savedInstanceState) {
+            String path = folder.path;
+            if (!TextUtils.isEmpty(mSubDirectoryTextView.getText())) {
+                path += mSubDirectoryTextView.getText();
+            }
```
This seems kind of hacky, would probably be cleaner if you used `getSavedSubDirectory()`.

jmintb replied:
I think I will just use replace, unless you think files would be better?

Nutomic replied:
Yes `File` is definitely better than using strings.

jmintb commented on src/main/java/com/nutomic/syncthingandroid/activities/ShareActivity.java:
```diff
@@ -123,7 +150,34 @@ protected void onCreate(Bundle savedInstanceState) {
+            String path = folder.path;
+            if (!TextUtils.isEmpty(mSubDirectoryTextView.getText())) {
+                path += mSubDirectoryTextView.getText();
+            }
```
I just spotted another problem, the folder path ends in a / and the subdirectory starts with a /. One of these should be removed.

Nutomic replied:
I think you can just use `File`s instead of strings for them, and it should take care of that automatically.

Nutomic commented:
One thing that I just noticed: if the root of the folder is selected, wouldnt it make more sense to display `/` instead of `browse`? Maybe we could have a field for the path, and a Button with `Select folder`. Also, I think the path should be shown with a trailing slash (right now it has a leading slash).

jmintb commented:
I think that it looks cleaner and less empty with browse instead of `/`. it would however be more informative to show the `/`, so maybe it does make more sense to show the slash.
I agree that there should be a trailing slash, should the leading slash then be removed?

I will start implementing the separate path field and browse button.

Nutomic commented:
Hmm, I guess it makes sense to show *something* if no folder is selected. But it could also be confusing, eg if you have a subfolder named `browse`. Yes I would only show a trailing slash for the path (that's how relative paths are usually shown on Linux).

jmintb commented:
Maybe we could show some sort of hint if no sub folder is selected, like "No sub folder is selected.".

jmintb commented:
Here is a quick implementation of the separate browse button and sub folder field. I am not sure whether or not the grey folder icon should be kept next to the textview. I havn't fixed the slashes yet.
What do you think?

![screenshot_20170829-215019](https://user-images.githubusercontent.com/10332534/29841408-9e7d01d0-8d05-11e7-9c7e-c0f223e84090.png)

![screenshot_20170829-215007](https://user-images.githubusercontent.com/10332534/29841416-a31261fe-8d05-11e7-983b-020ee72b8bf4.png)

Nutomic commented:
Yes that looks better. I would definitely keep the folder icon. For the button, I would put it to the right of the path TextView, and maybe just use the default grey background?

jmintb commented:
Yes, the default background sounds like a good idea.

jmintb commented:
There is now a new problem with the UI. When the keyboard is open, then the cancel and save buttons are pushed up, so they are behind the browse button. Any idea of how to deal with this?
I am not sure if we should put them on top of the browse button or if they should be kept below it, but then be hidden behind the keyboard. 
Maybe we could hide the sub directory UI elements while the keyboard is open. 
On phones with a larger screen and higher resolution I do not think this will be an issue, but on my 5 inch 720p display it certainly is. 

Sorry for the long delay, I have just been really busy in the past couple of weeks.

Nutomic commented:
I think the best option is if you just make the screen scrollable (eg wrap the root layout in a ScrollView). And no worries, just take your time ;)

jmintb commented:
Making the screen scrollable, introduced other issues relating to the position of the cancel and save button, so I replaced the relative layout with a constraint layout which solved these issues. What do you think?
![screenshot_20170920-222231](https://user-images.githubusercontent.com/10332534/30736770-679ba306-9f84-11e7-87f3-1be282b0a957.png)
This below image shows the ShareActivity with the keyboard open, where I have scrolled down to reveal the cancel and share button.
![screenshot_20170920-222242](https://user-images.githubusercontent.com/10332534/30736784-742008ba-9f84-11e7-887c-09cfbb12d78f.png)

Nutomic commented on src/main/java/com/nutomic/syncthingandroid/activities/ShareActivity.java:
```diff
@@ -272,4 +341,40 @@ protected void onPostExecute(Boolean isError) {
+                folderDirectory += "/";
+            }
+
+            String subDirectory = data.getStringExtra(FolderPickerActivity.EXTRA_RESULT_DIRECTORY) + "/";
```
Instead of messing around with path formatting manually, you can just use [`file.toURI().normalize().getPath()`](https://stackoverflow.com/a/21849556/1837158) to get the properly formatted path. You can put that in a method in `Util` and call it on FolderPickerActivity:200 and ShareActivity:361.

Nutomic commented:
The layout breaks on smaller screens (tested with Nexus 5, buttons are overlapping). So you will still need a ScrollView in that case. 

![screenshot_20170925-135457](https://user-images.githubusercontent.com/1044450/30792936-34e17464-a1f9-11e7-8db3-aa586065d175.png)

jmintb commented:
The view works fine on small screens now, at least from my testing.

Nutomic commented:
Thanks :)
